clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -1043,7 +1043,14 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
             /*ExcludeAspects=*/{"fp64"}));
         MPM.addPass(SYCLPropagateJointMatrixUsagePass());
       });
-    else if (LangOpts.SYCLIsHost && !LangOpts.SYCLESIMDBuildHostCode)
+      PB.registerOptimizerEarlyEPCallback([](ModulePassManager &MPM,
```

Comment:
What is the reason to use `registerOptimizerEarlyEPCallback` instead of `registerPipelineStartEPCallback`?

Reply:
As mentioned in https://github.com/intel/llvm/pull/16183/files/43526397fa2b4c559666fdfd6f302bfba7409255#r1857546652, we need to make sure this pass run after *AlwaysIniner*, moving to` PipelineStart` is what was tried in https://github.com/intel/llvm/pull/16183, it is before alwaysinliner and too early.

Reply:
@jsji, could you please extend the comment at lines 1049-1050 with the requirement to run _AlwaysInliner_ pass before _SYCLLowerWGLocalMemoryPass_?

Reply:
For sure.